### ISSUE TITLE
Webpack 5 throws "Target module of reexport from '...' is not part of the concatenation"

### ISSUE THREAD

**Issue by scinos** [webpack/webpack#12071]
# Bug report

Working on the migration of https://github.com/Automattic/wp-calypso to Webpack 5 I found an error that I can't figure out:

```
ERROR in ./node_modules/@wordpress/element/build-module/index.js + 1 modules
Target module of reexport from './node_modules/@wordpress/element/build-module/index.js' is not part of the concatenation
```

I tried to do a minimal reproduction case in https://github.com/scinos/webpack5-concat-reexport-error/, but unfortunately is not as minimal as I'd like :( I know having a reproduction case that depends on external npm dependencies is far from ideal, but I wasn't able to simplify it further.

What I know:
- Removing the `alias` in wepback config fixes the problem. Important: when resolved by Webpack, `@wordpress/data` is loaded from `node_modules/@wordpress/data/build-module`, aliasing it to `node_modules/@wordpress/data/build` shows the problem. AFAIK the only difference is `build-module` uses ESM and `build` uses CJS.

- It hits https://github.com/webpack/webpack/blob/master/lib/optimize/ConcatenatedModule.js#L413, trying to find the export for `Children` (which I believe it is reexported by https://github.com/WordPress/gutenberg/blob/master/packages/element/src/index.js#L2).

**What is the current behavior?**

Webpack throws
```
ERROR in ./node_modules/@wordpress/element/build-module/index.js + 1 modules
Target module of reexport from './node_modules/@wordpress/element/build-module/index.js' is not part of the concatenation
```

**If the current behavior is a bug, please provide the steps to reproduce.**
1. Checkout the reproduction repo
2. Run `yarn`
3. Run `yarn webpack`

**What is the expected behavior?**

Webpack compiles the project without errors

**Other relevant information:**
webpack version:  5.8.0
Node.js version: 12.18.4
Operating System: macOS 10.15.7

**Comment by alexander-akait** [comment]
Sorry, without reproducible test repo we can't help and we even don't know your code and configuration, please use stackoverflow for questions in this case

**Comment by scinos** [comment]
@alexander-akait The test repo is linked in the issue report, https://github.com/scinos/webpack5-concat-reexport-error/

**Comment by comoser** [comment]
@scinos I've recently tried to upgrade a project to Webpack 5.6 coming from 4.x.x.
And running the webpack command to build the project I was getting the `Target module of reexport from '...' is not part of the concatenation`, and the module was namely Material-UI.

Then I researched a bit more, and found this issue: https://github.com/webpack/webpack/issues/12058

The issue doesn't say much, but the sentence `Since 5.5.0 ...` made me try to upgrade to a lower version, namely 5.4.
Now the build passes. Not saying it's a solution, since the problem lies most likely with updates that specific libraries like Material UI need to do for more recent versions of webpack, but it may unblock your development until a solution is found.
